lib/vtls/rustls.c

```diff
@@ -203,45 +200,72 @@ rustls_send(struct connectdata *conn, int sockindex, const void *plainbuf,
   struct ssl_backend_data *const backend = connssl->backend;
   struct rustls_client_session *const session = backend->session;
   curl_socket_t sockfd = conn->sock[sockindex];
-  ssize_t plainwritten = 0;
-  ssize_t tlswritten = 0;
+  ssize_t n = 0;
+  size_t plainwritten = 0;
+  size_t tlslen = 0;
+  size_t tlswritten = 0;
   uint8_t tlsbuf[2048];
-  int result = 0;
 
   infof(data, "rustls_send of %d bytes\n", plainlen);
 
-  plainwritten = rustls_client_session_write(session, plainbuf, plainlen);
-  if(plainwritten == 0) {
-    failf(data, "rustls_send: EOF in write");
-    *err = CURLE_WRITE_ERROR;
-    return -1;
-  }
-  else if(plainwritten < 0) {
-    failf(data, "rustls_send: error in write");
-    *err = CURLE_WRITE_ERROR;
-    return -1;
-  }
+  if(plainlen > 0) {
+    n = rustls_client_session_write(session, plainbuf, plainlen);
+    if(n == 0) {
+      failf(data, "rustls_send: EOF in write");
+      *err = CURLE_WRITE_ERROR;
+      return -1;
+    }
+    else if(n < 0) {
+      failf(data, "rustls_send: error in write");
+      *err = CURLE_WRITE_ERROR;
+      return -1;
+    }
 
-  tlswritten = rustls_client_session_write_tls(
-      session, tlsbuf, sizeof(tlsbuf));
-  if(tlswritten == 0) {
-    failf(data, "rustls_send: EOF in write_tls");
-    *err = CURLE_WRITE_ERROR;
-    return -1;
-  }
-  else if(tlswritten < 0) {
-    failf(data, "rustls_send: error in write_tls");
-    *err = CURLE_WRITE_ERROR;
-    return -1;
+    DEBUGASSERT(n > 0);
+    plainwritten = n;
   }
 
+  while(rustls_client_session_wants_write(session)) {
+    n = rustls_client_session_write_tls(
+        session, tlsbuf, sizeof(tlsbuf));
+    if(n == 0) {
+      failf(data, "rustls_send: EOF in write_tls");
+      *err = CURLE_WRITE_ERROR;
+      return -1;
+    }
+    else if(n < 0) {
+      failf(data, "rustls_send: error in write_tls");
+      *err = CURLE_WRITE_ERROR;
+      return -1;
+    }
+
+    DEBUGASSERT(n > 0);
+    tlslen = n;
+    tlswritten = 0;
+
+    while(tlswritten < tlslen) {
+      n = swrite(sockfd, tlsbuf + tlswritten, tlslen - tlswritten);
+      if(n < 0) {
+        if(SOCKERRNO == EAGAIN || SOCKERRNO == EWOULDBLOCK) {
+          /* Since recv is called from poll, there should be room to
+            write at least some bytes before hitting EAGAIN. */
+          infof(data, "rustls_send: EAGAIN after %ld bytes\n", tlswritten);
+          DEBUGASSERT(tlswritten > 0);
+          break;
+        }
+        perror("writing to socket");
```

Comment:
Using `perror()` for sockets on Windows does not print the correct error. Add something like:
```c
static void Curl_perror (const char *what)
{
  char buf[300];
  fprintf (stderr, "%s: %s", what, Curl_strerror(SOCKERRNO, buf, sizeof(buf)));
}
```
at top. @bagder Should something like this be in `strerror.c`?

Reply:
It shouldn't be there